### ISSUE TITLE
Vim9: rules for disallowing shadowing of imported script are inconsistent and unpredictable

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#9563]
From [`:help :import`](https://vimhelp.org/vim9.txt.html#%3Aimport):

   > Then you can use "that.EXPORTED_CONST", "that.someValue", etc.  You are free
   > to choose the name "that".  Use something that will be recognized as referring
   > to the imported script.  Avoid command names and builtin function names,
   > because the name will shadow them.

There are several issues with this.

First, contrary to what the help suggests, the shadowing is not disallowed only for builtin functions; it's disallowed for any function, including user-defined ones:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/Map.vim')

import 'Map.vim'
def Map()
enddef
```
    E1236: Cannot use Map itself, it is imported

Here, an error is given as soon as we try to define a `Map()` function, because `Map` is already used as the name of an imported script.  This shows that shadowing is also disallowed for user-defined functions.

---

Second, the shadowing is not only disallowed to commands and functions.  It's also disallowed for variables:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/map.vim')

import 'map.vim'
var map: any
```
    E1213: Redefining imported item "map"

Currently, this is not documented.

---

Third, the rules regarding shadowing are inconsistent.  For functions, it's disallowed at the script-level:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/map.vim')

import 'map.vim'
echo range(3)->map((_, v) => v)
```
    E1236: Cannot use map itself, it is imported

But not in compiled code:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/map.vim')

import 'map.vim'
def Func()
    echo range(3)->map((_, v) => v * 2)
enddef
Func()
```
    [0, 2, 4]

For variables, things are yet again different.  Shadowing is always disallowed, including in compiled code:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/map.vim')

import 'map.vim'
def Func()
    var map: any
enddef
defcompile
```
    E1054: Variable already declared in the script: map

This is confusing.

I think it is impossible for the name of an imported script to shadow the name of a function or command.  But if I'm wrong, and we really must give an error, then, to be consistent, an error should *always* be given; preferably at compile time.

---

Fourth, when an error is given for a global function or a user-defined command, it's not given early enough:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/Func.vim')
lines =<< trim END
    vim9script
    def g:Func()
    enddef
END
lines->writefile('/tmp/t.vim')

import 'Func.vim'
source /tmp/t.vim
```
    no error

No error is given, and yet, the latter code is definitely broken, because calling `Func()` will always fail:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/Func.vim')
lines =<< trim END
    vim9script
    def g:Func()
    enddef
END
lines->writefile('/tmp/t.vim')

import 'Func.vim'
source /tmp/t.vim
Func()
```
    E1236: Cannot use Func itself, it is imported

Even more confusing, for a script-local function, an error is correctly given earlier (i.e. when the function is defined, not when it's called):
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/Func.vim')

import 'Func.vim'
def Func()
enddef
```
    E1236: Cannot use Func itself, it is imported

A similar issue affects user-defined commands:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/Cmd.vim')

import 'Cmd.vim'
command Cmd echomsg 'Cmd'
```
    no error

An error should be given as soon as `:Cmd` is defined, because executing it later will always fail:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
    enddef
END
lines->writefile(dir .. '/Cmd.vim')

import 'Cmd.vim'
command Cmd echomsg 'Cmd'
Cmd
```
    E1060: Expected dot after name: Cmd

Reversing the order between `:import` and `:command` makes no difference.

---

Fifth, I can see how the name of an imported script could shadow the name of a variable, but how could it shadow the name of a command/function?  There is no ambiguity with a command/function.

When using the name of an imported script, it must be followed by a dot.  But the name of a command cannot be followed by a dot.  It's either used alone on a line:

    edit

or followed by a bang:

    edit!
        ^

or followed by a whitespace:

    edit /path/to/file
        ^

Same thing for the name of a function: it must be followed by an open parenthesis:

    writefile([], 'file')
             ^

Although, since a function name can also be used as a funcref, I can think of a few other punctuation characters:

    # closing paren
    list->map(() => Ref)
                       ^

    # closing bracket
    var list = [Ref]
                   ^

    # comma
    var Partial = function(Ref, [0])
                              ^

    # hyphen
    Ref->Func()
       ^

    # bar
    var name = Ref| echo 'do something else'
                  ^

I might forget some characters, but I can't think of any context where the name of a function would be *immediately* followed by a dot.

Besides, in compiled code, Vim has currently no problem with us using the same name for an imported script as a builtin function:
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir->delete('rf')
dir ..= '/import'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    export def Item()
        echomsg 'from Item()'
    enddef
END
lines->writefile(dir .. '/map.vim')

import 'map.vim'
def Func()
    echo range(3)->map((_, v) => v * 2)
enddef
Func()
map.Item()
```
    [0, 2, 4]
    from Item()

This suggests that there is no technical reason for the limitation documented at [`:help :import`](https://vimhelp.org/vim9.txt.html#%3Aimport), and that we could remove it at the script level.  If it is possible, then I think we should do it.  Otherwise, whenever we want to use a *new* function or Ex command (new in the sense that we didn't already use it at least once in the script), then there is a risk of collision, which will require that either:

   - we rename our script (which is brittle because it could break other scripts which still import it under the old name)
   - we use the `as` keyword and import it under a new alias

Then, we need to rename all references to the imported script in the current script.  That's cumbersome and annoying.

Also, don't forget that:

   - the more scripts we import, the bigger the risk of collisions
   - we can use the import mechanism for autoload scripts, which significantly increases the number of scripts we can import
   - not even counting user-defined commands/functions, there are already more than a thousand builtin commands and functions (and this number can only grow over time):

         $ vim -Nu NONE +'vim9 echo len(getcompletion("", "command") + getcompletion("", "function"))'
         1126

     That's a lot of possible collisions.

Worse, a plugin author has no control over the user-defined commands which will be installed on the user's side.  Nor do they have control over their global functions.  If we can't make Vim smarter and let it infer that `Map` is a script in `Map.Item`, and a command in `Map` alone, then at least we should consider disallowing any name starting with an uppercase for an imported script.  Because even though a plugin author can test whether importing a script under the name `map` is OK on their machine, they can't do the same with the name `Map`.  That is, even if it works for them, it could still fail on a different machine.

**Comment by lacygoill** [comment]
> Besides, in compiled code, Vim has currently no problem with us using the same name for an imported script as a builtin function:

The snippet was wrong.  I fixed the code.

> I think it is impossible for the name of an imported script to shadow the name of a function or command.

I meant the opposite: I think it is impossible for the name of a function or command to shadow the name of an imported script.

> Fifth, I can see how the name of an imported script could shadow the name of a variable, but how could it shadow the name of a command/function? There is no ambiguity with a command/function.

Again, I meant the opposite.  I can see how the name of a variable can shadow the name of an imported script, but I can't see how a command or function could do the same.

---

TLDR: I think we still need to disallow a variable to shadow an imported script (which is not documented), but we should not give any error when calling a function or executing a command whose name matches the name of an imported script, because there is no need to, and because it creates too many issues.

**Comment by brammool** [comment]
> From [`:help :import`](https://vimhelp.org/vim9.txt.html#%3Aimport):
> 
>    > Then you can use "that.EXPORTED_CONST", "that.someValue", etc.  You are free
>    > to choose the name "that".  Use something that will be recognized as referring
>    > to the imported script.  Avoid command names and builtin function names,
>    > because the name will shadow them.
> 
> There are several issues with this.
> 
> First, contrary to what the help suggests, the shadowing is not disallowed only for builtin functions; it's disallowed for any function, including user-defined ones:
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/Map.vim')
> 
> import 'Map.vim'
> def Map()
> enddef
> ```
>     E1236: Cannot use Map itself, it is imported
> 
> Here, an error is given as soon as we try to define a `Map()`
> function, because `Map` is already used as the name of an imported
> script.  This shows that shadowing is also disallowed for user-defined
> functions.

That is not really what I would call shadowing.  You are using the same
name for two things.  The error message is confusing though.  I can fix
that.

> ---
> 
> Second, the shadowing is not only disallowed to commands and functions.  It's also disallowed for variables:
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/map.vim')
> 
> import 'map.vim'
> var map: any
> ```
>     E1213: Redefining imported item "map"
> 
> Currently, this is not documented.

Isn't it obvious that defining a variable with a name that is already
used for something is not allowed?  The error gives a more specific hint
about what the name is already used for, but you would get something
similar when the name is already used for a function or variable.

> ---
> 
> Third, the rules regarding shadowing are inconsistent.  For functions, it's disallowed at the script-level:
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/map.vim')
> 
> import 'map.vim'
> echo range(3)->map((_, v) => v)
> ```
>     E1236: Cannot use map itself, it is imported

This is the case that the help warns for: shadowing a builtin function
name can cause problems.

> But not in compiled code:
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/map.vim')
> 
> import 'map.vim'
> def Func()
>     echo range(3)->map((_, v) => v * 2)
> enddef
> Func()
> ```
>     [0, 2, 4]

That should probably also give an error, since "map" may mean one of two
things, and since we have allowed the import to shadow a builtin
variable, it should also apply here.

> For variables, things are yet again different.  Shadowing is always disallowed, including in compiled code:
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/map.vim')
> 
> import 'map.vim'
> def Func()
>     var map: any
> enddef
> defcompile
> ```
>     E1054: Variable already declared in the script: map
> 
> This is confusing.

This is the correct one.

> I think it is impossible for the name of an imported script to shadow
> the name of a function or command.  But if I'm wrong, and we really
> must give an error, then, to be consistent, an error should *always*
> be given; preferably at compile time.

The reason to allow shadowing is that after the script has beeen
written to import "somename", the builtin function "somename" could be
added, which would break the script.  Since we want the freedom to add
any builtin function, we don't want to break any existing scripts.


> Fourth, when an error is given for a global function or a user-defined
> command, it's not given early enough:
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/Func.vim')
> lines =<< trim END
>     vim9script
>     def g:Func()
>     enddef
> END
> lines->writefile('/tmp/t.vim')
> 
> import 'Func.vim'
> source /tmp/t.vim
> ```
>     no error
> 
> No error is given, and yet, the latter code is definitely broken,
> because calling `Func()` will always fail:
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/Func.vim')
> lines =<< trim END
>     vim9script
>     def g:Func()
>     enddef
> END
> lines->writefile('/tmp/t.vim')
> 
> import 'Func.vim'
> source /tmp/t.vim
> Func()
> ```
>     E1236: Cannot use Func itself, it is imported

This is calling the script-local Func, if you leave out the source
command it fails as well.  The script defines a global function, which
we know will only be used after any script-local function.

> Even more confusing, for a script-local function, an error is
> correctly given earlier (i.e. when the function is defined, not when
> it's called):
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/Func.vim')
> 
> import 'Func.vim'
> def Func()
> enddef
> ```
>     E1236: Cannot use Func itself, it is imported

That looks like the same as the very first example.

> A similar issue affects user-defined commands:
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/Cmd.vim')
> 
> import 'Cmd.vim'
> command Cmd echomsg 'Cmd'
> ```
>     no error
> 
> An error should be given as soon as `:Cmd` is defined, because executing it later will always fail:
> ```vim
> vim9script
> var dir = '/tmp/.vim'
> &runtimepath = dir
> dir->delete('rf')
> dir ..= '/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>     enddef
> END
> lines->writefile(dir .. '/Cmd.vim')
> 
> import 'Cmd.vim'
> command Cmd echomsg 'Cmd'
> Cmd
> ```
>     E1060: Expected dot after name: Cmd

It only fails when executed in the script.  If you use it elsewhere
there is no error.  And using it outside the script is the most common.
Therefore I don't think we should disallow this.

> Reversing the order between `:import` and `:command` makes no difference.
> 
> ---
> 
> Fifth, I can see how the name of an imported script could shadow the
> name of a variable, but how could it shadow the name of a
> command/function?  There is no ambiguity with a command/function.

Making a difference between a variable and a function sounds a bit
arbitrary though.
 
> When using the name of an imported script, it must be followed by a
> dot.  But the name of a command cannot be followed by a dot.  It's
> either used alone on a line:
> 
>     edit
> 
> or followed by a bang:
> 
>     edit!
>         ^
> 
> or followed by a whitespace:
> 
>     edit /path/to/file
>         ^
> 
> Same thing for the name of a function: it must be followed by an open
> parenthesis:
> 
>     writefile([], 'file')
>              ^

Allowing to use the import name without the dot for other things might
be very confusing.  A typo can easily lead to hard to find errors.
It is quite easy to use another name for the import.

> Although, since a function name can also be used as a funcref, I can
> think of a few other punctuation characters:
> 
>     # closing paren
>     list->map(() => Ref)
>                        ^
> 
>     # closing bracket
>     var list = [Ref]
>                    ^
> 
>     # comma
>     var Partial = function(Ref, [0])
>                               ^
> 
>     # hyphen
>     Ref->Func()
>        ^
> 
>     # bar
>     var name = Ref| echo 'do something else'
>                   ^
> 
> I might forget some characters, but I can't think of any context where
> the name of a function would be *immediately* followed by a dot.

One of the earlier examples was close to a case where it can get
confusing:

	import "name.vim"
	var name: {key: 'value'}
	echo name.key

Now "name.key" could be from the import or from the variable.  Therefore
we cannot allow for a variable to have the same name as an import.  And
as mentioned above, using a different rule for functions is a bit
unexpected.

> Besides, in compiled code, Vim has currently no problem with us using
> the same name for an imported script as a builtin function:
> ```vim
> vim9script
> delete('/tmp/.vim', 'rf')
> var dir = '/tmp/.vim/import'
> dir->mkdir('p')
> var lines =<< trim END
>     vim9script
>     export def Item()
>         echomsg 'from Item()'
>     enddef
> END
> lines->writefile(dir .. '/map.vim')
> &runtimepath = dir
> 
> import 'map.vim'
> def Func()
>     echo range(3)->map((_, v) => v * 2)
> enddef
> Func()
> map.Item()
> ```
>     [0, 2, 4]
>     from Item()
> 
> This suggests that there is no technical reason for the limitation
> documented at [`:help
> :import`](https://vimhelp.org/vim9.txt.html#%3Aimport), and that we
> could remove it at the script level.  If it is possible, then I think
> we should do it.  Otherwise, whenever we want to use a *new* function
> or Ex command (new in the sense that we didn't already use it at least
> once in the script), then there is a risk of collision, which will
> require that either:
> 
>    - we rename our script (which is brittle because it could break other scripts which still import it under the old name)
>    - we use the `as` keyword and import it under a new alias

Exactly, it's very easy to use a name that avoids name collisions.
While problems caused by using the same name in two different ways can
be hard to find.

> Then, we need to rename all references to the imported script in the
> current script.  That's cumbersome and annoying.

I disagree, this would hardly ever happen, and if it does the name
should be easy to find.  A global substitute probably works (unless you
use a word that might also appear in a comment perhaps.

> Also, don't forget that:
> 
>    - the more scripts we import, the bigger the risk of collisions
>    - we can use the import mechanism for autoload scripts, which
>    significantly increases the number of scripts we can import
>    - not even counting user-defined commands/functions, there are
>    already more than a thousand builtin commands and functions (and
>    this number can only grow over time):
> 
>          $ vim -Nu NONE +'vim9 echo len(getcompletion("", "command") + getcompletion("", "function"))'
>          1126
> 
>      That's a lot of possible collisions.
> 
> Worse, a plugin author has no control over the user-defined commands
> which will be installed on the user's side.  Nor do they have control
> over their global functions.  If we can't make Vim smarter and let it
> infer that `Map` is a script in `Map.Item`, and a command in `Map`
> alone, then at least we should consider disallowing any name starting
> with an uppercase for an imported script.  Because even though a
> plugin author can test whether importing a script under the name `map`
> is OK on their machine, they can't do the same with the name `Map`.
> That is, even if it works for them, it could still fail on a different
> machine.

Picking a good name is of course the solution to all this.  There are no
rules for that though.  Just be smart to not pick a name that is likely
to cause problems.

Disallowing an uppercase name might help with user defined functions,
but not with builtin functions.  I don't think it is a restriction that
is generally helpful.

-- 
GUEST:        He's killed the best man!
SECOND GUEST: (holding a limp WOMAN) He's killed my auntie.
FATHER:       No, please!  This is supposed to be a happy occasion!  Let's
              not bicker and argue about who killed who ...
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> Isn't it obvious that defining a variable with a name that is already used for something is not allowed?

It's probably obvious, but if we mention functions and commands at `:help :import`:

   > Then you can use "that.EXPORTED_CONST", "that.someValue", etc.  You are free
   > to choose the name "that".  Use something that will be recognized as referring
   > to the imported script.  Avoid command names and builtin function names,
   > because the name will shadow them.

then it seems weird to omit variables:
```diff
diff --git a/runtime/doc/vim9.txt b/runtime/doc/vim9.txt
index bbdc2bdf4..6d2da5907 100644
--- a/runtime/doc/vim9.txt
+++ b/runtime/doc/vim9.txt
@@ -1452,7 +1452,7 @@ In case the name is long or ambiguous, another name can be specified: >
 
 Then you can use "that.EXPORTED_CONST", "that.someValue", etc.  You are free
 to choose the name "that".  Use something that will be recognized as referring
-to the imported script.  Avoid command names and builtin function names,
+to the imported script.  Avoid names of commands, functions and variables,
 because the name will shadow them.
 
 In case the dot in the name is undesired, a local reference can be made for a
```
---

> Making a difference between a variable and a function sounds a bit
> arbitrary though.

It's not arbitrary.  We have control over the names of the variables in our script; we don't have any control over the names of the commands and global functions which are installed on the user side.

> And as mentioned above, using a different rule for functions is a bit
> unexpected.

Using a different rule could be easily justified, and could make the code more reliable, and give fewer useless errors.  When the feature was implemented, the very first error I got was:

    E1236: Cannot use map itself, it is imported

Which was completely useless, because my code was fine.

   > Disallowing an uppercase name might help with user defined functions,
   > but not with builtin functions.  I don't think it is a restriction that
   > is generally helpful.

It would be helpful, because at least it would give the guarantee that if the code works on one machine, it will work on any machine.

**Comment by lacygoill** [comment]
> It would be helpful, because at least it would give the guarantee that if the code works on one machine, it will work on any machine.

Never mind.  I was confused.  It should work on the user's side even if they have a command or function with the same name as an imported script.  There would be an issue if we wanted to call/execute it, but that doesn't make sense.  We can't call/execute something which we don't even know exist.

I still think it would be nice to allow `map()` even if we've imported `map.vim`, but I can live with it being disallowed.

**Comment by brammool** [comment]
> > Isn't it obvious that defining a variable with a name that is
> > already used for something is not allowed?
> 
> It's probably obvious, but if we mention functions and commands at
> `:help :import`:
> 
>    > Then you can use "that.EXPORTED_CONST", "that.someValue", etc.  You are free
>    > to choose the name "that".  Use something that will be recognized as referring
>    > to the imported script.  Avoid command names and builtin function names,
>    > because the name will shadow them.
> 
> then it seems weird to omit variables:

This refers to Vim commands and functions.  Thus if you would call the
import "edit" then you cannot use the ":edit" command.
As I mentioned before, we cannot forbid that, because then adding a
command or function later might break existing scripts.

If you define user commands, functions or variables in the same script,
that is not shadowing, that is redefining an item with a name that is
already in use.  This is what I call obvious: if you name things, you
can only use the name for one thing.

It is true that when the import starts with an uppercase letter you can
also shadow global user commands and functions.  I'll add a remark about
that in the help.

> > Making a difference between a variable and a function sounds a bit
> > arbitrary though.
> 
> It's not arbitrary.  We have control over the names of the variables
> in our script; we don't have any control over the names of the
> commands and global functions which are installed on the user side.

But those are not a problem at all, only script-local items are.
If you import something with the name "Foo" then obviously you cannot
use something global called "Foo".

Ah, for functions there is one workaround by prefixing "g:".  Same for
variables.  You can't do that for user commands though, but these aren't
used that much in a script.

> > And as mentioned above, using a different rule for functions is a bit
> > unexpected.
> 
> Using a different rule could be easily justified, and could make the
> code more reliable, and give fewer useless errors.  When the feature
> was implemented, the very first error I got was:
> 
>     E1236: Cannot use map itself, it is imported
> 
> Which was completely useless, because my code was fine.

I disagree. "map" is an import and you assume that from the following
"(" Vim will figure out that it can't be an import.  But what when using
"foobar(", which is not a builtin function, should there then be an
error for "no function called foobar" or "using import as a function"?
That shows the dilemma.

>    > Disallowing an uppercase name might help with user defined functions,
>    > but not with builtin functions.  I don't think it is a restriction that
>    > is generally helpful.
> 
> It would be helpful, because at least it would give the guarantee that
> if the code works on one machine, it will work on any machine.

That is another thing.  We can solve that be requiring that the file
name matches case, even though on MS-Windows case would be ignored.
There is another problem: two files that only differ in case would work
on Linux but not on MS-Windows.  But we can also just blame whoever did
this, it is a very common problem when sharing files between Linux and
MS-Windows.  Anyway, using ":source" has the same problem, requiring
matching case for ":import" would be unexpected.

-- 
There are three kinds of people: Those who can count & those who can't.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
